Translation of .qgs project data

signedav opened the pull request:
*Implementation of [QEP #90](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/90)*

### Summary from QEP:
> QGIS is used in different environments also with different localization requirements. Plugins can be translated using Qt possibilities (tr('Hi') + pylupdate + lrelease), the QGIS interface itself is also translated.
Plugins often ship with preconfigured project files. It would be great to be able to ship these with translation information. This QEP proposes to do this in a very similar way that the Qt ecosystem does this always.

### Implemented workflow
#### 1. Creation of translationfile:
- You are in a .qgs project file with a base language (e.g. English: testproject.qgs)
- On Project Properties \* you can *Create a .ts translation file* for the selected language (e.g. English)
<image>
The ts file is stored like this <qgs-project-name>_<selected_language_code>.ts (e.g. testproject_de.ts)

\* do you think, this is the right place for it?
- With e.g. Qt Linguist you can translate the .ts file and generate the .qm file (or with `lrealease`) 

#### 2. Translation of projectfile:
- You open a project (e.g. testproject.qgs) with an individual base language (e.g. German)
- In case you have a qm-file that fits to the project and the language (e.g. testproject_de.qm) the mentioned below texts are translated to German.
- The translated project is stored after loading with the language postfix (e.g. testproject_de.qgs)

### Translatable objects
- [x] layer names
- [x] layer group names
- [ ] layer sub group names
- [x] relation names
- [ ] project name
- [x] attribute alias (if no alias, then the field translation)
- [x] forms group box titles
- [x] forms sub group box titles
- [ ] labels
- [ ] expressions
- [ ] action titles
- [ ] constraint messages
- [ ] composer texts (?)
- [ ] ValueRelation -> value column
- [ ] ValueMap -> values

Additional things to do (beside the missing translatable objects):
- Extend the ts file instead of overwriting. Because otherwise it has to be translated completely again, when added e.g. a field.

signedav committed:
generate ts file

with the structure for it - not finished

signedav committed:
translator basics

basics work - translation of layername, relations
fields and aliases have a problem

signedav committed:
tests, project store and other

not yet working all translations
project handling can be improved
project store after translation working
and fix some errors
tests not ready yet

signedav committed:
gui implementation

signedav committed:
working translation of project from qm file

Translated objects and tests included:
- [x] layer names
- [x] layer group names
- [x] relation names
- [x] attribute alias (if no alias, then the field translation)
- [x] forms group box titles
- [x] forms sub group box titles

included restructured with QTranslator object.

Updated test-data used for the test.

nyalldawson commented:
I have some issues with this approach. Firstly, it's reintroducing QgsProject::instance calls to core, so its going to cause issues for server (immediately) and make the eventual removal of QgsProject::instance much more complex (4.0).

Secondly, I think it's important to ensure that the implementation works with non file project storage, because the most likely use case for this feature will be larger organisations and projects were they are more likely to be using database project storage.

To me this is a "proceed with extreme caution" pr - we are potentially setting API decisions here which will make long term maintenance more complex.

nyalldawson commented:
Just to make my above comment more constructive, I'd suggest that if translation only occurs on project load, you could use QgsReadWriteContext to handle the translation and avoid the QgsProject:: instance calls.

signedav committed:
QgsProjectTranslator

implementation of abstract class QgsProjectTranslator

signedav committed:
setProjectTranslator

...in QgsReadWriteContext creation
implementation of projectTranslator where needed

use of DefaultTranslator to have default translation funtion derived from QgsProjectTranslator

beautify code

signedav committed:
use of smartpointer mTranslator

for better handling if loaded (or in our case existent) or not...

signedav commented:
Instead of using QgsProject::inctance(), for "passing" project-functions with `QgsReadWriteContext` I use an abstract class `QgsProjectTranslator` where `translate()` is derived by `QgsProject` (or with DefaultTranslator to pass back the original value in case of no project-instance).

> Secondly, I think it's important to ensure that the implementation works with non file project storage, because the most likely use case for this feature will be larger organisations and projects were they are more likely to be using database project storage.

We decided to make it only for the file project storage as a first step, because this is what's apparently most requested and described in the QEP. I hope this is okay for you.

signedav committed:
docstrings etc.

m-kuhn commented on src/app/qgsprojectproperties.cpp:
```diff
@@ -830,6 +830,29 @@ QgsProjectProperties::QgsProjectProperties( QgsMapCanvas *mapCanvas, QWidget *pa
   connect( titleEdit, &QLineEdit::textChanged, mMetadataWidget, &QgsMetadataWidget::setTitle );
 
+  //fill ts language checkbox
+  //fill ts language checkbox
```
Double comment?

m-kuhn commented on src/app/qgsprojectproperties.cpp:
```diff
@@ -830,6 +830,29 @@ QgsProjectProperties::QgsProjectProperties( QgsMapCanvas *mapCanvas, QWidget *pa
+    cbtsLocale->addItem( QIcon( QString( ":/images/flags/%1.svg" ).arg( l ) ), displayName, l );
+  }
+
+  cbtsLocale->addItem( QIcon( QString( ":/images/flags/%1.svg" ).arg( QStringLiteral( "en_US" ) ) ), QLocale( QStringLiteral( "en_US" ) ).nativeLanguageName(), QStringLiteral( "en_US" ) );
```
`QStringLiteral( ":/images" ...`

m-kuhn commented on src/app/qgsprojectproperties.h:
```diff
@@ -92,6 +92,9 @@ class APP_EXPORT QgsProjectProperties : public QgsOptionsDialogBase, private Ui:
     void scaleItemChanged( QListWidgetItem *changedScaleItem );
 
+    //! generate the ts file with the locale selected in the checkbox
+    void onGenerateTsFileButton();
```
I think this method can be const?

m-kuhn commented on src/core/qgsapplication.h:
```diff
@@ -732,6 +733,13 @@ class CORE_EXPORT QgsApplication : public QApplication
+    /**
+     * Emits the signal to collect all the strings of .qgs to be included in ts file
+     *
+     * \since QGIS 3.2
```
Watch out with the version

m-kuhn commented on src/core/qgslayerdefinition.cpp:
```diff
@@ -51,6 +51,7 @@ bool QgsLayerDefinition::loadLayerDefinition( const QString &path, QgsProject *p
 
   QgsReadWriteContext context;
   context.setPathResolver( QgsPathResolver( path ) );
+  context.setProjectTranslator( project );
```
I guess this it for opening a `.qlr` file. I'm not sure if we want to apply the translations from the currently loaded project or rather in the future have QGIS look for `.qm` translations next to the `.qlr` file itself... Opinions?

signedav replied:
I'm not sure. When I think about it, I guess I prefer not to translate it when loading the `.qlr`, because there is most possibly no translation available anyway (because the ts file has been created without it...). Or do you think someone does that on purpose?
I like your idea to have a `.qm` file for the `.qlr` itself...

m-kuhn commented on src/core/qgsproject.cpp:
```diff
@@ -441,6 +442,66 @@ void QgsProject::setPresetHomePath( const QString &path )
+void QgsProject::registerTranslatableObjects( QgsTranslationContext *translationContext )
+{
+  //register layers
+  for ( QgsLayerTreeLayer *layer : mRootGroup->findLayers() )
```
copy before loop

m-kuhn commented on src/core/qgsproject.cpp:
```diff
@@ -441,6 +442,66 @@ void QgsProject::setPresetHomePath( const QString &path )
 
+void QgsProject::registerTranslatableContainers( QgsTranslationContext *translationContext, QgsAttributeEditorContainer *parent, const QString &layerId )
+{
+  QList<QgsAttributeEditorElement *> elements = parent->children();
```
const

m-kuhn commented on src/core/qgsproject.cpp:
```diff
@@ -441,6 +442,66 @@ void QgsProject::setPresetHomePath( const QString &path )
+  }
+
+  //register layergroups
+  for ( const QgsLayerTreeGroup *groupLayer : mRootGroup->findGroups() )
```
copy

signedav commented on src/core/qgsproject.cpp:
```diff
@@ -1219,6 +1315,23 @@ bool QgsProject::readProjectFile( const QString &filename )
+    }
+    else
+    {
+      QgsDebugMsg( "Error saving translated project with locale prefix " + newFileName );
```
I agree - it's quite an important information.

signedav committed:
sip files and some improvements

like make a list copy before go through for loop and log messages

signedav committed:
spelling

m-kuhn commented on src/core/qgseditformconfig.h:
```diff
@@ -288,7 +288,7 @@ class CORE_EXPORT QgsEditFormConfig
      * Deserialize drag and drop designer elements.
      */
-    QgsAttributeEditorElement *attributeEditorElementFromDomElement( QDomElement &elem, QgsAttributeEditorElement *parent );
+    QgsAttributeEditorElement *attributeEditorElementFromDomElement( QDomElement &elem, QgsAttributeEditorElement *parent, const QString &layerId = nullptr );
```
= QString()

m-kuhn commented:
Failing sip checks because of changes introduced in sip during the lifetime of a PR start to become a bit painful.

@3nids what's your opinion on removing generated sip files from the source tree and making them build-time only?

m-kuhn commented on src/core/qgstranslationcontext.cpp:
```diff
@@ -0,0 +1,98 @@
+/***************************************************************************
+  qgstranslationcontext.cpp - %{Cpp:License:ClassName}
```
There's a placeholder left here

m-kuhn commented on src/core/qgstranslationcontext.cpp:
```diff
@@ -0,0 +1,98 @@
+ ---------------------
+ begin                : 23.5.2018
+ copyright            : (C) 2018 by david
+ email                : [your-email-here]
```
And here

m-kuhn committed:
Add missing sip bindings

3nids committed:
add missing sip file

3nids committed:
fix SIP test

signedav committed:
fixes

3nids commented:
failing of the code_layout build is now unrelated.
there are some warnings when building astyle and dash.orfeo seems off at the moment, so I cannot look at the warning (not echoed on Travis).
So from a code layout / SIP point of view, LGTM

3nids commented:
oops
https://dash.orfeo-toolbox.org/viewBuildError.php?type=1&buildid=328791
layerId is not documented

signedav committed:
missing parameter documentation

signedav commented:
> Instead of using QgsProject::inctance(), for "passing" project-functions with QgsReadWriteContext I use QgsProjectTranslator where translate() is derived by QgsProject (or with DefaultTranslator to pass back the original value in case of no project-instance).

> We decided to make it only for the file project storage as a first step, because this is what's apparently most requested and described in the QEP.

Are you okay with that as well @nyalldawson ?

signedav commented:
Actually in case of .qgz the .qm file *needs* to be in the qgz-archive. I guess it's what's expected and it's even for end users not that hard to add it there after translating and releasing it manually, because on most system you can drag and drop files into it.

On the other hand, the generated .ts files are not in the qgz-archive but next to the .qgz file. And this is what I prefer because with the .ts file you have to do something (open it with Qt Linguist e.g. what wouldn't that user friendly if you have to export it first from the .qgz).

What we could think about is, that it's not only using the .qm file on the .qgz when it's in the archive, but even when it's next to it. But I'm not sure about it, because AFAIK it's not usual to have file next to it on .qgz projects.

m-kuhn commented:
Ok for everyone if I merge this tomorrow?

nyalldawson commented on src/core/layertree/qgslayertreegroup.cpp:
```diff
@@ -252,12 +252,24 @@ QgsLayerTreeGroup *QgsLayerTreeGroup::findGroup( const QString &name )
   return nullptr;
 }
 
+QList<QgsLayerTreeGroup *> QgsLayerTreeGroup::findGroups() const
```
Can you add a unit test for this function please?

nyalldawson commented on src/core/layertree/qgslayertreenode.cpp:
```diff
@@ -65,6 +65,7 @@ QgsLayerTreeNode *QgsLayerTreeNode::readXml( QDomElement &element, const QgsProj
   if ( project )
     resolver = project->pathResolver();
   context.setPathResolver( resolver );
+  context.setProjectTranslator( ( QgsProject * )project );
```
c++ style cast please

nyalldawson commented on src/core/layout/qgscompositionconverter.cpp:
```diff
@@ -1193,6 +1193,7 @@ bool QgsCompositionConverter::readLegendXml( QgsLayoutItemLegend *layoutItem, co
     pathResolver = project->pathResolver();
   QgsReadWriteContext context;
   context.setPathResolver( pathResolver );
+  context.setProjectTranslator( ( QgsProject * )project );
```
dynamic/static cast

nyalldawson commented on src/core/qgsapplication.h:
```diff
@@ -755,6 +763,13 @@ class CORE_EXPORT QgsApplication : public QApplication
     void nullRepresentationChanged();
 
+    /**
+     * requests vor strings of .qgs to be included in ts file
```
vor -> for, "requests" -> "Requests"

Actually nicer is

    Emitted when project strings which require translation are being collected for inclusion in a .ts file.

    In order to register translatable strings, connect to this signal and register the strings within the specified \a translationContext.

nyalldawson commented on src/core/qgseditformconfig.cpp:
```diff
@@ -518,13 +518,13 @@ void QgsEditFormConfig::writeXml( QDomNode &node, const QgsReadWriteContext &con
   if ( elem.tagName() == QLatin1String( "attributeEditorContainer" ) )
   {
-    QgsAttributeEditorContainer *container = new QgsAttributeEditorContainer( elem.attribute( QStringLiteral( "name" ) ), parent );
+    QgsAttributeEditorContainer *container = new QgsAttributeEditorContainer( QgsProject::instance()->translate( QStringLiteral( "project:layers:%1:formcontainers" ).arg( layerId ), elem.attribute( QStringLiteral( "name" ) ) ), parent );
```
Remove QgsProject::instance() call

nyalldawson commented on src/core/qgslayerdefinition.cpp:
```diff
@@ -298,6 +299,7 @@ QList<QgsMapLayer *> QgsLayerDefinition::loadLayerDefinitionLayers( const QStrin
 
   QgsReadWriteContext context;
   context.setPathResolver( QgsPathResolver( qlrfile ) );
+  //no projecttranslator defined here
```
"project translator"

nyalldawson commented on src/core/qgsproject.cpp:
```diff
@@ -441,6 +442,69 @@ void QgsProject::setPresetHomePath( const QString &path )
+
+      translationContext->registerTranslation( QStringLiteral( "project:layers:%1:formcontainers" ).arg( layerId ), container->name() );
+
+      if ( container->children().size() > 0 )
```
if ( !container->children().empty() )

nyalldawson commented on src/core/qgsproject.cpp:
```diff
@@ -921,6 +1008,16 @@ bool QgsProject::readProjectFile( const QString &filename )
 
+  QgsSettings settings;
+
+  QString localeFileName = QStringLiteral( "%1_%2" ).arg( QFileInfo( projectFile.fileName() ).baseName(), settings.value( QStringLiteral( "locale/userLocale" ), "" ).toString() );
```
"" -> QString()

nyalldawson commented on src/core/qgsproject.h:
```diff
@@ -969,7 +998,7 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      * Emitted when a project is being read.
      */
-    void readProject( const QDomDocument & );
+    void readProject( const QDomDocument &, QgsReadWriteContext &context );
```
I'm unsure - is this an api break? It could cause issues with existing connections

nyalldawson commented on src/core/qgsproject.h:
```diff
@@ -1269,6 +1298,26 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
     void setPresetHomePath( const QString &path );
 
+    /**
+     * Registers the translatable containers into the translation context
```
rekursive -> recursive. Also punctuation is missing here.

nyalldawson commented on src/core/qgsproject.h:
```diff
@@ -1269,6 +1298,26 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+    void registerTranslatableContainers( QgsTranslationContext *translationContext, QgsAttributeEditorContainer *parent, const QString &layerId );
+
+    /**
+     * Registers the translatable objects into the translation context
```
"Registers translatable objects into the translation context for inclusion in ts files."

nyalldawson commented on src/core/qgsprojecttranslator.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * \ingroup core
+ * This abstract class is to call translate() for project data from wherever QgsReadWriteContext is available.
```
The language here is unclear, can you reword? (I realise it's a translation, but it's difficult to understand)

nyalldawson commented on src/core/qgsprojecttranslator.h:
```diff
@@ -0,0 +1,46 @@
+  public:
+
+    /**
+     * This method needs to be reimplemented in all classes which implement this interface
```
That's evident from the =0 :wink: . Better describe what it actually does here instead.

nyalldawson commented on src/core/qgsreadwritecontext.h:
```diff
@@ -99,16 +100,36 @@ class CORE_EXPORT QgsReadWriteContext
+    const QgsProjectTranslator *projectTranslator( ) const { return mProjectTranslator; }
+
+    /**
+     * Sets the project translator
```
Make a note about ownership/lifetime of projectTranslator here.

nyalldawson commented on src/core/qgsreadwritecontext.h:
```diff
@@ -99,16 +100,36 @@ class CORE_EXPORT QgsReadWriteContext
     QList<ReadWriteMessage> mMessages;
     QStringList mCategories = QStringList();
-
+    QgsProjectTranslator *mProjectTranslator;
```
initialize to nullptr

nyalldawson commented on src/core/qgstranslationcontext.cpp:
```diff
@@ -0,0 +1,98 @@
+
+#include "qgssettings.h"
+
+QgsTranslationContext::QgsTranslationContext()
```
= default in header

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+/**
+ * \ingroup core
+ * \class QgsTranslationContext
+ * \brief used for the collecting of strings of .qgs to be translated and writing of ts file
```
Capitalization

"Used for collecting strings from projects for translation and creation of ts files."

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+
+class CORE_EXPORT QgsTranslationContext
+{
+    struct TranslatableObject
```
Missing dox

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+{
+    struct TranslatableObject
+    {
+      QString context;
```
Missing dox on these members too

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+    QgsProject *project() const;
+
+    /**
+     * Sets the \a project where the translation need to be done for
```
Sets the \a project being translated.

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+    QString fileName() const;
+
+    /**
+     * Sets the \a name of the TS file
```
\a fileName (or rename fileName -> name )

nyalldawson commented on src/core/qgstranslationcontext.h:
```diff
@@ -0,0 +1,95 @@
+    void setFileName( const QString &fileName );
+
+    /**
+     * Registers the \a string to be translated
```
"string" argument does not exist

signedav committed:
layermaptest and fixes

after feedback - mostly doc stuff or code style

signedav committed:
merge error

m-kuhn commented:
Thanks for the review and fixing!
Good for everyone?

signedav commented:
Thanks guys @3nids @nyalldawson @m-kuhn :+1:

m-kuhn commented:
Hey there, we just tested this with transifex integration. And good news: it works :tada: 

Some notes:

* I think by default, the translation should be empty (set to `<translation type="unfinished"></translation>`)
* Make it clear that the language supplied in the configuration dialog is the target language (and the source language is the current project language). This needs to be made clear in the project properties dialog and the API.
  Question: Do we even need that? At least on transifex you can upload one source file (e.g. in English) and all target files will be produced. Not sure how that's handled with QtLinguist
* For unfinished translations, it will just translate to an empty string. In this case it should show the untranslated source string. At the moment, downloading an untranslated or partially translated .ts from transifex, converting to a .qm and loading that will make the layer tree completely without text.

Concerning translated values:

* The Value Relation -> Value field would be very valuable to translate too (lookup from `value_en` to `value_de`)

But anyway, good job!
